### DISCUSSION TITLE
如果要判断不同编译工具，怎么处理，特别是在嵌入式系统中，不同编译工具，对应的代码可能会有不一样。

### DISCUSSION THREAD

**Discussion by niurenyige** [xmake-io/xmake#3647] [General]
set(HC32F030_SDK_VER HC32F030_DDL_Rev1.3.1)
if(${CMAKE_C_COMPILER_ID} MATCHES "ARMCC")
    #target_sources(mcu PRIVATE common/startup_hc32f005.c)
    target_sources(mcu PRIVATE  common/startup_hc32f030_armcc.s)
    set(LINKER_SCRIPT  ${CMAKE_CURRENT_LIST_DIR}/common/armcc_hc32f030.sct)

    tos_compile_options($<$<COMPILE_LANGUAGE:C>:--c99 -c --cpu=Cortex-M0+ -g -O2 --apcs=interwork --split_sections > )
    tos_compile_definitions($<$<COMPILE_LANGUAGE:C>:__MICROLIB>)

    tos_compile_options( $<$<COMPILE_LANGUAGE:ASM>: --predefine "__MICROLIB SETL {TRUE}" >)
    tos_compile_options( $<$<COMPILE_LANGUAGE:ASM>:--cpu Cortex-M0+ -g --apcs=interwork >)

    tos_ld_options(--cpu=Cortex-M0+ --library_type=microlib --strict --scatter ${LINKER_SCRIPT}  )
    tos_ld_options(--keep=*Handler )
else()
    message(STATUS "GNU GCC COMPILER")
    tos_compile_options(-mcpu=cortex-m0plus  -mthumb)
    tos_ld_options(-mcpu=cortex-m0plus  -mthumb)
    set(LINKER_SCRIPT "${CMAKE_CURRENT_LIST_DIR}/common/gcc_hc32f030.ld")
    tos_ld_options( "-Wl,--gc-sections -T ${LINKER_SCRIPT}")

    target_sources(mcu PRIVATE common/startup_hc32f030.c)
endif()

这个是我以前用CMAKE 写的，想用XMAKE 来做。如果有is_toolchain("gcc") is_toolchain("armcc") 这样就好处理。

**Comment by waruqi** [comment]
https://github.com/xmake-io/xmake/blob/f11dad73c0d087f43ea5e016208f0a4a05334fd6/xmake/rules/c%2B%2B/modules/modules_support/common.lua#L191

**Reply by ruiyuanlu** [reply]
请问在 xmake.lua 中如何写呢？if target:has_tool("msvc") then xxx 会报错
